recipes/Libri-Light/self-supervised-learning/make_librilight_csv.py

```diff
@@ -0,0 +1,168 @@
+
+    Arguments
+    ---------
+    data_path : str
+        Path to the Libri-Light split after vad
+    csv_file_folder: str
+        Path to the output folder of generated csv files
+    max_length: float
+        The max length (seconds) for each prepared audio clip
+    n_processes: int
+        Number of parallel processes
+    """
+    print("Processing each subfolder of this split")
+    os.makedirs(csv_file_folder, exist_ok=True)
+    os.makedirs(f"{csv_file_folder}/tmp", exist_ok=True)
+    list_dir = pathlib.Path(data_path)
+    tasks = []
+    for x in list_dir.iterdir():
+        tasks.append((x, f"{csv_file_folder}/tmp", max_length))
+    with multiprocessing.Pool(processes=n_processes) as pool:
+        for _ in tqdm.tqdm(
+            pool.imap_unordered(make_csv_for_each, tasks), total=len(tasks)
+        ):
+            pass
+
+
+def merge_csv_files(
+    csv_file_folder: str,
+) -> None:
+    """Merges the csv files for each subfolder into one single file
+
+    Arguments
+    ---------
+    csv_file_folder: str
+        Path to the output folder of generated csv files
+    """
+    print("Merging the csvs of each subfolder into one csv")
+    file_list = [str(x) for x in Path(f"{csv_file_folder}/tmp").glob("*.csv")]
+    output_file = f"{csv_file_folder}/train.csv"
+    fieldnames = ["ID", "duration", "wav"]
+
+    with open(output_file, mode="a", newline="", encoding="utf-8") as outfile:
+        csv_writer = csv.writer(
+            outfile, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL
+        )
+        csv_writer.writerow(fieldnames)
+        for file_path in tqdm.tqdm(file_list):
+            with open(file_path, mode="r", encoding="utf-8") as infile:
+                reader = csv.reader(infile)
+
+                # filter out bad rows
+                for row in reader:
+                    if len(row) == 3 and os.path.exists(row[-1]):
+                        new_row = [row[-1], row[1], row[2]]
+                        csv_writer.writerow(new_row)
+                    else:
+                        print(f"bad row {row}")
+
+    import shutil
+
+    shutil.rmtree(f"{csv_file_folder}/tmp")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--input_dir",
+        type=str,
+        default=None,
+        help="Path to the Libri-Light split after vad",
+        required=True,
+    )
+    parser.add_argument(
+        "--output_dir",
+        type=str,
+        default=None,
+        help="Path to the output folder of generated csv files",
+        required=True,
+    )
+    parser.add_argument(
+        "--max_length",
+        type=float,
+        default=20.2,
+        help="The max length for each prepared audio clip" "(default is 20.2)",
+    )
+    parser.add_argument(
+        "--n_processes",
+        type=int,
+        default=32,
+        help="Number of parallel processes",
+    )
+    args = parser.parse_args()
+    return args
```

Comment:
@TParcollet do you have an opinion on that? Usually the data prep is being done directly in the train.py ,BUT,  given how large the dataset is, I would say that 99% of the folks will prefer to first process the dataset, and then, train models (which means a clear way of executing only the data prep). In this regard, I would say we could stick to argparse for this recipe, but I'd like to get the goat opinion on this design choice. thanks

Reply:
This is an excellent question as I have been dealing with recipes with 10k+ hours recently and i've had this exact same issue. I resorted in adding a yaml argument to just do the dataprep. 

I think that the key here is to find the potential issue that it may cause. If one starts a training with a very long data prep on a single process, personally, I don't think it is a problem and it will work just fine.  Problem is ... these very long data prep imply that people will want to use multiple GPUs, hence DDP, hence time out on the dataprep. There is also the cluster occupation issue, as it is inefficient to occupy multiple GPUs for a few hours.

I think that what @shucongzhang proposes here makes sense, and we should convert Gigaspeech to do the same. However, we would also need to discuss more about that with people like @pplantinga, @mravanelli and @asu.

Reply:
Hi @Adel-Moumen How are you. Thank you so much for the quick and detailed comments! I really appreciate it! For the script, I've rewritten it with SpeechBrain utils functions. For having a separate  data preparation, I think the reasons are

1. Before the SpeechBrain data preparation, the user needs to use external VAD to pre-process the data, which is already a separate step. 
2. There is no dev set for LibriLight. Thus, the dev.csv should be taken from somewhere else,  which is also already a separate data preparation. 
3. As you said, the data is huge and prepare it first before training may be a good option. 

I'm also discussing with Titouan about this. I'll update this PR after there's a conclusion about how should we do the data preparation for this dataset. 

Thank you again!

Reply:
> Hi @Adel-Moumen How are you. Thank you so much for the quick and detailed comments! I really appreciate it! For the script, I've rewritten it with SpeechBrain utils functions. For having a separate data preparation, I think the reasons are
> 
>     1. Before the SpeechBrain data preparation, the user needs to use external VAD to pre-process the data, which is already a separate step.
> 
>     2. There is no dev set for LibriLight. Thus, the dev.csv should be taken from somewhere else,  which is also already a separate data preparation.
> 
>     3. As you said, the data is huge and prepare it first before training may be a good option.
> 
> 
> I'm also discussing with Titouan about this. I'll update this PR after there's a conclusion about how should we do the data preparation for this dataset.
> 
> Thank you again!

Hi! 

>     1. Before the SpeechBrain data preparation, the user needs to use external VAD to pre-process the data, which is already a separate step.

can you please elaborate on this? I didn't knew you had to run VAD on top of Libri-Light. According to the official paper, they said that they already ran a VAD no?

>     2. There is no dev set for LibriLight. Thus, the dev.csv should be taken from somewhere else,  which is also already a separate data preparation.

For the dev set, I guess we can directly in `train.py` or something like that take `N%` of training set as dev set? 

Let me know the outcome of the discussion. 

Do you plan to add a recipe e.g. best-rq or wav2vec? Woudl be nice to have that with this PR.